### PR TITLE
IDEMPIERE-3334: Adding support for Readonly Logic, Set Pressed Logic …

### PR DESCRIPTION
**Pull request by dpansheriya**
…on custom toolbar button

### CODE REVIEW

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADWindowToolbar.java [hunk 1]
```diff
@@ -898,6 +898,8 @@ public void updateToolbarAccess() {
 		ADWindow adwindow = ADWindow.findADWindow(this);
 
 		dynamicDisplay();
+		pressedLogic();
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADWindowToolbar.java:901]: I think it is better to call both logic method at the end of dynamicDisplay instead. With this, we can make both *logic method private and don't have to hunt for places where dynamicDisplay is call.

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADWindowToolbar.java (hunk 1 above)
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADWindowToolbar.java]: @hengsin  Please note that I made changes as suggested.

**Review by hengsin — CHANGES_REQUESTED** [review]
please rename migration script to current timestamp instead of using timestamp from 2017.

**Comment by CarlosRuiz-globalqss** [comment]
The oracle script has two problems:

- ALTER TABLE ... VARCHAR2(2000) ->  must be written as VARCHAR2(2000 CHAR)
- there are lines starting with the character @ - these throw errors when running with sqlplus - this was solved in [IDEMPIERE-4481](https://idempiere.atlassian.net/browse/IDEMPIERE-4481) - and the script here is using the old description of the element

**Comment by hengsin** [comment]
hi @dpansheriya , any update on this ?

**Comment by dpansheriya** [comment]
> * sqlplus

It got completely missed in holiday season here. Let me work on it.

**Comment by dpansheriya** [comment]
> * VARCHAR2(2000 CHAR)

Please note that done all and pushed.

**Comment by dpansheriya** [comment]
> please rename migration script to current timestamp instead of using timestamp from 2017.

This is done.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java [hunk 2]
```diff
@@ -104,7 +107,77 @@ public void dynamicDisplay(boolean forceValidation) {
...
+			isSQL = true;
+		}
+
+		logic = Env.parseContext(Env.getCtx(), windowNo, tabNo, logic, false);
+
+		if (isSQL) {
+			int result = DB.getSQLValue(null, logic);
+
+			if (result == 1)
+				isValid = true;
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java:173]: This is inconsistent with the use of SQL logic in other part of the application. Preferably, should use Evaluator.parseSQLLogic instead.

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java (hunk 2 above)
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java]: Changed to use Evaluator.parseSQLLogic

**Review by hengsin — CHANGES_REQUESTED** [review]
Not visible (not implemented ?) for Window > Tab > Toolbar Button.

**Comment by dpansheriya** [comment]
> 

@hengsin  may be I am not understanding your questions. What I understand is you are asking that display logic not implemented. But display logic field already exists in master branch.  Can you please clarify if I misunderstood your point?

**Comment by hengsin** [comment]
hi @dpansheriya , no, I means at "Window, Tab and Field", the Readonly Logic and Pressed Logic is not shown. Is that by design or just overlook ?

**Comment by dpansheriya** [comment]
> hi @dpansheriya , no, I means at "Window, Tab and Field", at the "Toolbar Button" tab, the Readonly Logic and Pressed Logic field is not shown. Is that by design or just overlook ?

Ahh, We have added Pressed Logic and Read Only logic on Toolbar window only. Not on Window->Tab->Toolbar button.

**Comment by hengsin** [comment]
hi @dpansheriya , that's using the same table (AD_ToolbarButton), shouldn't that be added there as well ?

**Comment by dpansheriya** [comment]
> hi @dpansheriya , that's using the same table (AD_ToolbarButton), shouldn't that be added there as well ?

I think yes, So let us add and test.

**Comment by dpansheriya** [comment]
> hi @dpansheriya , that's using the same table (AD_ToolbarButton), shouldn't that be added there as well ?

@hengsin  please note this support added.

**Comment by hengsin** [comment]
[pr1244.patch.txt](https://github.com/idempiere/idempiere/files/10049193/pr1244.patch.txt)
- context will be evaluated in evaluateLogic. current approach cause evaluation exception if context is not parseable.

**Review by hengsin — CHANGES_REQUESTED** [review]
hi @dpansheriya , please see attached patch above.

**Comment by dpansheriya** [comment]
> hi @dpansheriya , please see attached patch above.

Applied

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java [hunk 3]
```diff
@@ -104,7 +106,70 @@ public void dynamicDisplay(boolean forceValidation) {
...
+	private boolean validateLogic(String logic, int tabNo) {
+		boolean isValid = false;
+
+		if (logic.startsWith("@SQL="))
+		{
+			isValid = Evaluator.parseSQLLogic(logic, Env.getCtx(), windowNo, tabNo, "");
+		}
+		else
+		{
+			logic = Env.parseContext(Env.getCtx(), windowNo, tabNo, logic, false);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java:166]: hi @dpansheriya , miss this in my previous patch, this line should be removed as well.

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java (hunk 3 above)
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java]: @hengsin  this is done.

**Review by hengsin — APPROVED** [review]
tested

**Pull request merged**